internal/impl/pinecone/output.go

```diff
@@ -0,0 +1,369 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pinecone
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/pinecone-io/go-pinecone/pinecone"
+	"github.com/redpanda-data/benthos/v4/public/bloblang"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+const (
+	poFieldBatching        = "batching"
+	poFieldHost            = "host"
+	poFieldAPIKey          = "api_key"
+	poFieldNamespace       = "namespace"
+	poFieldID              = "id"
+	poFieldOp              = "operation"
+	poFieldVectorMapping   = "vector_mapping"
+	poFieldMetadataMapping = "metadata_mapping"
+)
+
+func outputSpec() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Version("0.1.0").
+		Categories("AI").
+		Summary("Inserts items into a Pinecone index.").
+		Description(service.OutputPerformanceDocs(true, true)).
+		Fields(
+			service.NewOutputMaxInFlightField(),
+			service.NewBatchPolicyField(poFieldBatching),
+			service.NewStringField(poFieldHost).
+				Description("The host for the Pinecone index."),
+			service.NewStringField(poFieldAPIKey).
+				Secret().
+				Description("The Pinecone api key."),
+			service.NewStringEnumField(poFieldOp, string(operationUpdate), string(operationUpsert), string(operationDelete)).
+				Default(string(operationUpsert)).
+				Description("The operation to perform against the Pinecone index."),
+			service.NewInterpolatedStringField(poFieldNamespace).
+				Default("").
+				Advanced().
+				Description("The namespace to write to - writes to the default namespace by default."),
+			service.NewInterpolatedStringField(poFieldID).
+				Description("The ID for the index entry in Pinecone."),
+			service.NewBloblangField(poFieldVectorMapping).
+				Optional().
+				Description("The mapping to extract out the vector from the document. The result must be a floating point array. Required if not a delete operation."),
```

Comment:
I don't think we have anywhere where we need a single type as the output. Happy to take suggestions for other ways to let folks map this.

Reply:
Yeah I think this is probably the best overall UX. Ideally we'd add a few examples here to the field of mappings like might make sense. Something like `root = this.foos.map_each(foo -> foo.blah)`, but we can put together some typical examples later.